src/viper/common.ml

```diff
@@ -4,3 +4,23 @@ exception Unsupported of Source.region * string
 let unsupported at sexp =
   raise (Unsupported (at, (Wasm.Sexpr.to_string 80 sexp)))
 
+let rec zip_with : f:('a -> 'b -> 'c) -> 'a list -> 'b list -> 'c list option =
+  fun ~f xs ys ->
+    match xs, ys with
+    | [], [] -> Some []
+    | x :: xs, y :: ys -> Option.bind (zip_with ~f xs ys) (fun xys -> Option.some @@ f x y :: xys)
+    | _ :: _, [] | [], _ :: _ -> None
+
+let zip_with_exn : f:('a -> 'b -> 'c) -> 'a list -> 'b list -> 'c list =
+  fun ~f xs ys ->
+    match zip_with ~f xs ys with
+    | Some xys -> xys
+    | None -> failwith "Expected lists to have the same length. %s"
+
+let zip : 'a list -> 'b list -> ('a * 'b) list option =
+  fun xs ys ->
+    zip_with ~f:(fun x y -> x, y) xs ys
+
+let zip_exn : 'a list -> 'b list -> ('a * 'b) list =
+  fun xs ys ->
+    zip_with_exn ~f:(fun x y -> x, y) xs ys
```

Comment:
I am not sure, but probably it would be more "in Ocaml style" to use `List.combine` and `List.fold_left2`.
In your case I think you need just `zip_with_ext` which is 
```ocaml
List.fold_left2 (func zs x y -> (x, y)::zs) [] xs ys
```

Reply:
I got used to the combinators from the `core` library. I didn't know that they have different names in the stdlib. Stripped these functions